### PR TITLE
Favicon support: way more work than anyone expected or considered

### PR DESCRIPTION
**Pull request by zmike**
Getting these down to the embedding API level required that I redo the bindings generator again, so this is more commits than anticipated.

@mbrubeck @Manishearth @pcwalton but NOT @larsbergstrom so don't even look at this.

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/6219)

### CODE REVIEW

File: ports/cef/interfaces/cef_drag_data.rs [hunk 1]
```diff
@@ -499,7 +499,7 @@ impl CefDragData {
   // Retrieve the list of file names that are being dragged into the browser
   // window.
   //
-  pub fn get_file_names(&self, names: Vec<String>) -> libc::c_int {
+  pub fn get_file_names(&self, names: &Vec<String>) -> libc::c_int {
```
Comment (pcwalton) [inline ports/cef/interfaces/cef_drag_data.rs]: Kinda weird to use `&Vec<String>` instead of `&[String]`, but OK I guess.

**Comment by Manishearth** [comment]
Review status: 3 of 26 files reviewed, 2 unresolved discussions, all commit checks successful.
Reviewed files:
- components/compositing/compositor_task.rs @ r5
- components/script/dom/htmlbodyelement.rs @ r5
- components/script/dom/htmllinkelement.rs @ r1

---

<sup>**[components/script/dom/htmllinkelement.rs, line 94 \[r6\]](https://reviewable.io:443/reviews/servo/servo/6219#-JqWoroQiJEssVDBlF-B)** ([raw file](https://github.com/servo/servo/blob/d02019ddd473debe8faae903c856fd9effe323b0/components/script/dom/htmllinkelement.rs#L94)):</sup>
won't `s.eq_ignore_ascii_case("shortcut icon")` never happen because of the split on spaces?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6219)

**Comment by Manishearth** [comment]
script LGTM modulo one issue

**Comment by zmike** [comment]
Review status: 3 of 26 files reviewed, 2 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/dom/htmllinkelement.rs, line 94 \[r6\]](https://reviewable.io:443/reviews/servo/servo/6219#-JqWoroQiJEssVDBlF-B)** ([raw file](https://github.com/servo/servo/blob/d02019ddd473debe8faae903c856fd9effe323b0/components/script/dom/htmllinkelement.rs#L94)):</sup>
Seems to work. Or at least I get favicons from these types.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6219)

File: ports/cef/interfaces/cef_drag_data.rs (hunk 1 above)
Comment (zmike) [inline ports/cef/interfaces/cef_drag_data.rs]: I made the smallest changes possible to make it work.

**Comment by Manishearth** [comment]
Review status: 3 of 26 files reviewed, 2 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/dom/htmllinkelement.rs, line 94 \[r6\]](https://reviewable.io:443/reviews/servo/servo/6219#-JqWoroQiJEssVDBlF-B)** ([raw file](https://github.com/servo/servo/blob/d02019ddd473debe8faae903c856fd9effe323b0/components/script/dom/htmllinkelement.rs#L94)):</sup>
"shortcut icon" will match "icon" when split on spaces :smile: 

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6219)

**Comment by zmike** [comment]
Review status: 3 of 26 files reviewed, 2 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/dom/htmllinkelement.rs, line 94 \[r6\]](https://reviewable.io:443/reviews/servo/servo/6219#-JqWoroQiJEssVDBlF-B)** ([raw file](https://github.com/servo/servo/blob/d02019ddd473debe8faae903c856fd9effe323b0/components/script/dom/htmllinkelement.rs#L94)):</sup>
Aha. Can I just have "icon" then since it would match both types?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6219)

**Comment by zmike** [comment]
@Manishearth ?

**Comment by Manishearth** [comment]
(I'm only going to review script)

---

Review status: 3 of 26 files reviewed, 2 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/dom/htmllinkelement.rs, line 94 \[r6\]](https://reviewable.io:443/reviews/servo/servo/6219#-JqWoroQiJEssVDBlF-B)** ([raw file](https://github.com/servo/servo/blob/d02019ddd473debe8faae903c856fd9effe323b0/components/script/dom/htmllinkelement.rs#L94)):</sup>
Yes

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/6219)

**Comment by metajack** [comment]
Assigning to @pcwalton for non-script parts.

**Comment by zmike** [comment]
Updated to address @Manishearth comment.

File: components/script/dom/htmllinkelement.rs [hunk 2]
```diff
@@ -85,6 +87,16 @@ fn is_stylesheet(value: &Option<String>) -> bool {
     }
 }
 
+fn is_favicon(value: &Option<String>) -> bool {
```
Comment (pcwalton) [inline components/script/dom/htmllinkelement.rs]: Is there a spec for this to cite?

**Comment by pcwalton** [comment]
Looks OK overall, but I'd like to know what spec, if any, this is following.

**Comment by zmike** [comment]
I'm using https://html.spec.whatwg.org/multipage/semantics.html#rel-icon in accordance with the use case in CEF, where only a list of icon URLs is passed to the application.

**Comment by pcwalton** [comment]
OK, could you add a link to that spec in the relevant parts in the DOM crate? Otherwise this looks OK.

**Comment by zmike** [comment]
Spec note added.

**Comment by pcwalton** [comment]
Squash and then r=me

**Pull request merged**